WalletWasabi/BitcoinCore/Monitoring/P2pReconnector.cs

```diff
@@ -12,47 +12,41 @@ public class P2pReconnector : PeriodicRunner
 		public P2pReconnector(TimeSpan period, P2pNode p2pNode) : base(period)
 		{
 			P2pNode = Guard.NotNull(nameof(p2pNode), p2pNode);
-			Success = new TaskCompletionSource<object>();
+			Success = new TaskCompletionSource<bool>();
 		}
 
 		public P2pNode P2pNode { get; set; }
-		private TaskCompletionSource<object> Success { get; }
+		private TaskCompletionSource<bool> Success { get; }
 
 		protected override async Task ActionAsync(CancellationToken cancel)
 		{
-			try
+			Logger.LogInfo("Trying to reconnect to P2P...");
+			if (await P2pNode.TryDisconnectAsync(cancel).ConfigureAwait(false))
 			{
-				Logger.LogInfo("Trying to reconnect to P2P...");
-				P2pNode.Disconnect();
 				await P2pNode.ConnectAsync(cancel).ConfigureAwait(false);
-			}
-			catch
-			{
-				P2pNode.Disconnect();
-				throw;
-			}
 
-			Logger.LogInfo("Successfully reconnected to P2P.");
-
-			Success.TrySetResult(null);
+				Logger.LogInfo("Successfully reconnected to P2P.");
+				Success.TrySetResult(true);
+			}
 		}
 
 		public async Task StartAndAwaitReconnectionAsync(CancellationToken cancel)
 		{
 			await StartAsync(cancel).ConfigureAwait(false);
-			using var ctr = cancel.Register(() => Success.SetResult(null));
+			using var ctr = cancel.Register(() => Success.SetResult(false));
 			await Success.Task.ConfigureAwait(false);
```

Comment:
When https://github.com/zkSNACKs/WalletWasabi/pull/4707/files#diff-04d1d7e67ec7f785cb1e52f4512613ab91efc98a9544a40c32fd9d189d3d1efcR24 keeps throwing exceptions, you'll be stuck here forever, right? Is that acceptable or not?

Reply:
Yes. If someone is concerned with it he can add a timeout here: https://github.com/molnard/WalletWasabi/blob/790f9c098280a50f818c5bd5cdb4ff53cff82eaf/WalletWasabi/BitcoinCore/P2pNode.cs#L98

In case of Wasabi disposal it will be canceled gracefully - so that is covered.